### DISCUSSION TITLE
Renew hook documentation

### DISCUSSION THREAD

**Discussion by dubidubno** [acmesh-official/acme.sh#4548] [Q&A] [answered]
Is there any documentation availale on how to write a renew hook? I can't seem to find anything on this other than https://github.com/acmesh-official/acme.sh/wiki/Using-pre-hook-post-hook-renew-hook-reloadcmd

Is it just a standalone shell script?  What parameters are available for the script? How does the script know which certificate got renewed?

**Comment by FarrelF** [comment]
Sorry for the late answer. Renew Hook is just a shell script that will be executed if you have successfully renewed your certificates, the renew hook script using your acme.sh certificate directory as a working directory, for example: `/home/<USERNAME>/.acme.sh/domain.com`

So if you want to use renew hook that will be executing the file which is located outside certificate directory, then you must using command that execute the script with complete script file path as a renew hook, for example: `sh /path/to/the/script.sh`

I personally recommend that you create a shell script file in your certificate directory after issuing a certificate, so you can just using `./your-script.sh`, `sh your-script.sh`, `bash your-script.sh`, `/usr/bin/env sh your-script.sh` or `/usr/bin/env bash your-script.sh` or whatever instead of mentioning full script file path like `sh /path/to/the/script.sh` in your renew hook.

Because renew hook using certificate directory as a working directory, you can using your certificate files (like: `domain.com.cer`, `domain.com.key`, `ca.cer` and `fullchain.cer` in `/home/<USERNAME>/.acme.sh/domain.com` path) without having to mention the full file path.

### Setting up Renew Hook 

For setting up a renew hook, while you want issuing a certificate, then you can add `--renew-hook command` argument.

Change `command` with shell command that you will execute while successfully renewed, if you want to execute a script file, you can just change `command` to something like `sh your-script.sh` and then you can create a shell script file in your certificate directory after that.

If you've already issued a certificate before, then you can just set up the renew hook with add `Le_RenewHook` option to your domain configuration file which is located at your certificate directory (For example: `/home/user/.acme.sh/domain.com/domain.com.conf`).

You can fill `Le_RenewHook` option with any shell command and don't forget that renew hook using your certificate directory as a working directory. After the command successfully executed, the `Le_RenewHook` command that you filled will be encoded into Base64 and changed to the following:

```plain
__ACME_BASE64__START_<BASE64_COMMAND>__ACME_BASE64__END_
```

The `<BASE64_COMMAND>` is your renew hook command which is encoded into Base64.

That's all a little information about renew hook feature in acme.sh this time, I hope it helps you.

**Comment by dubidubno** [comment]
So the domain name in question and paths to the certificate files are not passed to the renew hook script?

**Reply by FarrelF** [reply, marked as answer]
In acme.sh code, there is a few lines that [export some variables](https://github.com/acmesh-official/acme.sh/blob/15ee036db17d13a8ec75275e0b7ef729a79dc9f2/acme.sh#L3629-L3634), including `CERT_PATH`, `CERT_KEY_PATH`, `CA_CERT_PATH`, `Le_Domain` + `DOMAIN_PATH` that you can try to insert it to your renew hook script.

About working directory, it seems that acme.sh using `cd` to `$DOMAIN_PATH`, that's how they're using certificate file path as a working directory.

**tl;dr:** All of these variables can display outputs perfectly without any errors while executing renew hook, even when you're using these variables in script file as a renew hook. So you can pass the domain folder name and certificate file path with exported variable to your renew hook script if needed.

To be honest, I just tried all of these variables in my renew hook script, with issuing some a new test/staging certificate using `/data/data/com.termux/files/usr/bin/env sh deploy.sh` as a renew hook and after issued successfully I make a `deploy.sh` file into certificate path (For example in this case: `/data/data/com.termux/files/home/.acme.sh/staging.domain.com/deploy.sh`) with following script commands:

```shell
#!/data/data/com.termux/files/usr/bin/env sh
echo $CERT_PATH
echo $CERT_KEY_PATH
echo $CA_CERT_PATH
echo $CERT_FULLCHAIN_PATH
echo $Le_Domain
echo $DOMAIN_PATH
```

**Note:** I tried this in Termux, so the `/data/data/com.termux/files/usr/bin/env` command is equivalent to `/usr/bin/env` and `/data/data/com.termux/files/home/.acme.sh` path is equivalent to `/home/<YOUR_USERNAME>/.acme.sh`

Yes, the above script are only to print these variables while succesfully renewed. Then, I forcefully renew the staging certificate with following command:

```shell
acme.sh --renew -d staging.domain.com --ecc --force --server googletest
```

I used Google Public CA Staging Server in this case to issue the staging certificate before, so I use `--server googletest` argument to prevent acme.sh switch ACME Server to production server of Google Public CA. The certificate was renewed successfully, the script was executed successfully and I got this following output:

```plain
/data/data/com.termux/files/home/.acme.sh/staging.domain.com_ecc/ssl-staging.domain.com.cer	# $CERT_PATH
/data/data/com.termux/files/home/.acme.sh/staging.domain.com_ecc/ssl-staging.domain.com.key	# $CERT_KEY_PATH
/data/data/com.termux/files/home/.acme.sh/staging.domain.com_ecc/ca.cer   # $CA_CERT_PATH
/data/data/com.termux/files/home/.acme.sh/staging.domain.com_ecc/fullchain.cer  # $CERT_FULLCHAIN_PATH
staging.domain.com_ecc		# $Le_Domain
/data/data/com.termux/files/home/.acme.sh/staging.domain.com_ecc	# $DOMAIN_PATH
```

As you can see above, the script outputs it perfectly without any errors. So, you can pass the domain folder name and certificate file path with exported variable to your renew hook script.

Sorry for my bad english 🙏
